### DISCUSSION TITLE
Problems with getting pipenv manager to function as expected (grouping, skipping deps, skipping specific deps)

### DISCUSSION THREAD

**Discussion by Atheuz** [renovatebot/renovate#16358] [Request Help]
I'm just starting out with Renovate and I have a Python project that's serving as an example of how to use Renovate with Python.
The project repo is in Azure, it uses pipenv for dependency management, renovate is running in a Docker container using the latest version (renovate/renovate:latest, as of 2022-07-01, seems to be "renovateVersion": "32.99.9").
The `renovate-config.js` for the build repo hosting the basic renovate config looks like this:

```
// renovate config file

module.exports = {
		timezone: 'Europe/Copenhagen',
		extends: [
			":dependencyDashboard",
			":semanticPrefixFixDepsChoreOthers",
			":ignoreModulesAndTests",
			":autodetectPinVersions",
			":prHourlyLimit2",
			":prConcurrentLimit10",
			"group:monorepos",
			"group:recommended",
			"workarounds:all",
			":prHourlyLimitNone"
		],
		azureWorkItemId: 123456789,
		endpoint: 'http://tfs/tfs/DefaultCollection',

		platform: 'azure',
		onboardingConfig: {
			extends: ['config:base'],
		},

		repositories: [
			'test/test-repo'
		],

		allowCustomCrateRegistries: true,
		allowScripts: true,
		exposeAllEnv: true,
		dryRun: true,
		printConfig: true,

		"python": {
			"registryUrls": ["https://artifactory/artifactory/api/pypi/python/simple/"]
		},
};
```

The `renovate.json` file for the target test-repo looks like this:

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "description": "Python dependencies and Docker images",
  "packageRules": [
    {
      "enabled": false,
      "groupName": "all-python-packages",
      "groupSlug": "all-python-packages",
      "matchDatasources": "python",
      "matchManagers": "pipenv"
    },
    {
      "enabled": false,
      "groupName": "all-major-python-packages",
      "groupSlug": "all-major-python-packages",
      "matchDatasources": "python",
      "matchManagers": "pipenv",
      "matchUpdateTypes": [
        "major"
      ]
    },
    {
      "enabled": false,
      "groupName": "all-python-dev-packages",
      "groupSlug": "all-python-dev-packages",
      "matchDatasources": "python",
      "matchDepTypes": [
        "dev-packages"
      ],
      "matchManagers": "pipenv"
    },
    {
      "excludePackageNames": [
        "mccabe"
      ],
      "groupName": "python-packages",
      "groupSlug": "python-packages",
      "lockFileMaintenance": {
        "enabled": true
      },
      "matchDatasources": "python",
      "matchDepTypes": [
        "packages"
      ],
      "matchManagers": "pipenv",
      "matchPackageNames": [
        "fastapi",
        "autodynatrace",
        "uvicorn",
        "hypercorn",
        "starlette-exporter",
        "requests",
        "psutil",
        "confluent-kafka",
        "protobuf"
      ],
      "matchUpdateTypes": [
        "minor",
        "patch"
      ]
    }
  ]
}
```

The Pipfile looks like this:

```
[[source]]
url = "https://artifactory/artifactory/api/pypi/python/simple"
verify_ssl = true
name = "artifactory-pypi"

[packages]
fastapi = "==0.78.0"
autodynatrace = "==1.0.84"
uvicorn = {version = ">=0.12.0,<0.18.0", extras=["standard"]}
hypercorn = {version = "==0.13.2", extras = ["trio"]}
starlette-exporter = "==0.13.0"
requests = ">=2.24.0,<3.0.0"
psutil = "==5.9.1"
confluent-kafka = "==1.9.0"
protobuf = "==3.20.1"

[dev-packages]
pre-commit = "==2.19.0"
pre-commit-hooks = "==4.3.0"
black = "==22.6.0"
isort = "==5.10.1"
mypy = "==0.961"
flake8 = "==4.0.1"
mccabe = "==0.6.1"
flake8-docstrings = "==1.6.0"
flake8-bugbear = "==22.6.22"
flake8-absolute-import = "==1.0.0.1"
pep8-naming = "==0.13.0"
flake8-print = "==5.0.0"
pytest = "==7.1.2"
pytest-cov = "==3.0.0"
pytest-env = "==0.6.2"
pytest-mock = "==3.7.0"
"punch.py" = "==2.0.0"
yamllint = "==1.26.3"
types-protobuf = "==3.19.22"
types-requests = "==2.28.0"
responses = "==0.21.0"

[requires]

[pipenv]
allow_prereleases = false

```

When the build pipeline is run, it outputs the config because I've set it to printConfig: true, and also set it to dryRun: true so it doesn't actually create any PRs (but the outcome is the same if I set dryRun to false). 
My expectation is that it skips the dev-packages, it skips major versions, it ignores updates to mccabe (if for some reason it doesn't skip dev-packages), it creates a single PR, and maybe that it updates the Pipfile.lock file too (not 100% sure about that).
But what actually happens is this:

```
...
          {
            "enabled": false,
            "groupName": "all-python-packages",
            "groupSlug": "all-python-packages",
            "matchDatasources": ["python"],
            "matchManagers": ["pipenv"]
          },
          {
            "enabled": false,
            "groupName": "all-major-python-packages",
            "groupSlug": "all-major-python-packages",
            "matchDatasources": ["python"],
            "matchManagers": ["pipenv"],
            "matchUpdateTypes": ["major"]
          },
          {
            "enabled": false,
            "groupName": "all-python-dev-packages",
            "groupSlug": "all-python-dev-packages",
            "matchDatasources": ["python"],
            "matchDepTypes": ["dev-packages"],
            "matchManagers": ["pipenv"]
          },
          {
            "excludePackageNames": ["mccabe"],
            "groupName": "python-packages",
            "groupSlug": "python-packages",
            "lockFileMaintenance": {"enabled": true},
            "matchDatasources": ["python"],
            "matchDepTypes": ["packages"],
            "matchManagers": ["pipenv"],
            "matchPackageNames": [
              "fastapi",
              "autodynatrace",
              "uvicorn",
              "hypercorn",
              "starlette-exporter",
              "requests",
              "psutil",
              "confluent-kafka",
              "protobuf"
            ],
            "matchUpdateTypes": ["minor", "patch"]
          },
...
WARN: findIssue() is not implemented (repository=test/test-repo)
INFO: Dependency extraction complete (repository=test/test-repo)
      "baseBranch": "master",
      "stats": {
        "managers": {
          "docker-compose": {"fileCount": 4, "depCount": 1},
          "dockerfile": {"fileCount": 4, "depCount": 6},
          "helm-values": {"fileCount": 1, "depCount": 2},
          "helmv3": {"fileCount": 1, "depCount": 1},
          "pipenv": {"fileCount": 1, "depCount": 35}
        },
        "total": {"fileCount": 11, "depCount": 45}
      }
(Use `node --trace-warnings ...` to show where the warning was created)
INFO: DRY-RUN: Would commit files to branch renovate/mccabe-0.x (repository=test/test-repo, branch=renovate/mccabe-0.x)
INFO: DRY-RUN: Would create PR: Update dependency mccabe to v0.7.0 (repository=test/test-repo, branch=renovate/mccabe-0.x)
WARN: artifactErrors (repository=test/test-repo, branch=renovate/mccabe-0.x)
      "artifactErrors": [
        {
          "lockFile": "Pipfile.lock",
          "stderr": "Command failed: pipenv lock\nLocking [dev-packages] dependencies...\n\r⠋\r Locking...Building requirements...\n\b\r⠙ Locking...\b\r⠹ Locking...\b\r⠸ Locking...\rResolving dependencies...\n\b\r⠼ Locking...\b\r⠴ ✘ Locking Failed! \n\nCRITICAL:pipenv.patched.notpip._internal.resolution.resolvelib.factory:Cannot install -r /tmp/pipenvj9e6ahzzrequirements/pipenv-ot6c2if4-constraints.txt (line 17) and mccabe==0.7.0 because these package versions have conflicting dependencies.\n[ResolutionFailure]:   File \"/opt/buildpack/tools/python/3.10.5/lib/python3.10/site-packages/pipenv/resolver.py\", line 822, in _main\n[ResolutionFailure]:       resolve_packages(\n[ResolutionFailure]:   File \"/opt/buildpack/tools/python/3.10.5/lib/python3.10/site-packages/pipenv/resolver.py\", line 771, in resolve_packages\n[ResolutionFailure]:       results, resolver = resolve(\n[ResolutionFailure]:   File \"/opt/buildpack/tools/python/3.10.5/lib/python3.10/site-packages/pipenv/resolver.py\", line 751, in resolve\n[ResolutionFailure]:       return resolve_deps(\n[ResolutionFailure]:   File \"/opt/buildpack/tools/python/3.10.5/lib/python3.10/site-packages/pipenv/utils/resolver.py\", line 1068, in resolve_deps\n[ResolutionFailure]:       results, hashes, markers_lookup, resolver, skipped = actually_resolve_deps(\n[ResolutionFailure]:   File \"/opt/buildpack/tools/python/3.10.5/lib/python3.10/site-packages/pipenv/utils/resolver.py\", line 862, in actually_resolve_deps\n[ResolutionFailure]:       resolver.resolve()\n[ResolutionFailure]:   File \"/opt/buildpack/tools/python/3.10.5/lib/python3.10/site-packages/pipenv/utils/resolver.py\", line 663, in resolve\n[ResolutionFailure]:       raise ResolutionFailure(message=str(e))\n[pipenv.exceptions.ResolutionFailure]: Warning: Your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.\n  You can use $ pipenv install --skip-lock to bypass this mechanism, then run $ pipenv graph to inspect the situation.\n  Hint: try $ pipenv lock --pre if it is a pre-release dependency.\nERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts\n\n"
        }
      ]
INFO: DRY-RUN: Would ensure lock file error comment in PR #0 (repository=test/test-repo, branch=renovate/mccabe-0.x)
INFO: DRY-RUN: Would commit files to branch renovate/pytest-mock-3.x (repository=test/test-repo, branch=renovate/pytest-mock-3.x)
INFO: DRY-RUN: Would create PR: Update dependency pytest-mock to v3.8.1 (repository=test/test-repo, branch=renovate/pytest-mock-3.x)
INFO: DRY-RUN: Would commit files to branch renovate/uvicorn-0.x (repository=test/test-repo, branch=renovate/uvicorn-0.x)
INFO: DRY-RUN: Would commit files to branch renovate/protobuf-4.x (repository=test/test-repo, branch=renovate/protobuf-4.x)
INFO: DRY-RUN: Would create PR: Update dependency protobuf to v4 (repository=test/test-repo, branch=renovate/protobuf-4.x)
INFO: DRY-RUN: Would ensure Dependency Dashboard (repository=test/test-repo)
      "title": "Dependency Dashboard"
INFO: Repository finished (repository=test/test-repo)
      "durationMs": 467511
##[section]Finishing: Run Renovate Bot on repos
```

1. It doesn't skip dev-packages (as can be seen with it trying to update pytest-mock and mccabe).
2. It doesn't skip major releases (as seen with it trying to update protobuf to v4).
3. It doesn't skip mccabe (as can be seen with it trying (and failing) to update mccabe to 0.7).
4. It wants to create multiple PRs.

Can someone see what I'm doing wrong, or if there is an issue with pipenv that I'm not aware of?

**Comment by rarkins** [comment]
If you have multiple problems,
Please reduce down to smaller reproduction with one problem at a time. Your config and example is too complex to easily review

**Comment by Atheuz** [comment]
@rarkins here's a simpler example that only includes the problem of not skipping major versions, is this more what you were thinking?

`Pipfile`: I've installed protobuf==3.20.0 (which is one version behind, 3.20.1, and a major version behind 4.21.2)
```
[[source]]
url = "https://artifactory/artifactory/api/pypi/python/simple"
verify_ssl = true
name = "artifactory-pypi"

[packages]
protobuf = "==3.20.0"

[dev-packages]

[requires]
```

`renovate.json`: We disable major in matchUpdateTypes, and target protobuf and minor, patch versions of protobuf.
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "description": "Python dependencies and Docker images",
  "packageRules": [
    {
      "enabled": false,
      "groupName": "all-major-python-packages",
      "groupSlug": "all-major-python-packages",
      "matchDatasources": ["python"],
      "matchManagers": ["pipenv"],
      "matchUpdateTypes": [
        "major"
      ]
    },
    {
      "enabled": true,
      "groupName": "python-packages",
      "groupSlug": "python-packages",
      "matchDatasources": ["python"],
      "lockFileMaintenance": {
        "enabled": true
      },
      "matchDepTypes": [
        "packages"
      ],
      "matchManagers": ["pipenv"],
      "matchPackageNames": [
        "protobuf"
      ],
      "matchUpdateTypes": [
        "minor",
        "patch"
      ]
    }
  ]
}
```

`renovate.config.js`: Then we target this new repo in `renovate.config.js` in the renovate build repository

```
// renovate config file

module.exports = {
		timezone: 'Europe/Copenhagen',
		extends: [
			":dependencyDashboard",
			":semanticPrefixFixDepsChoreOthers",
			":ignoreModulesAndTests",
			":autodetectPinVersions",
			":prHourlyLimit2",
			":prConcurrentLimit10",
			"group:monorepos",
			"group:recommended",
			"workarounds:all",
			":prHourlyLimitNone"
		],
		azureWorkItemId: 123456,
		endpoint: 'http://tfs/tfs/DefaultCollection',

		platform: 'azure',
		onboardingConfig: {
			extends: ['config:base'],
		},

		repositories: [
			'test/renovate-test-repo'
		],

		allowCustomCrateRegistries: true,
		allowScripts: true,
		exposeAllEnv: true,
		dryRun: false,
		printConfig: true,

		"python": {
			"registryUrls": ["https://artifacts.sys.dom/artifactory/api/pypi/python/simple/"]
		}
};

```

Then running the Renovate Bot build, the relevant output is:

```
...
          {
            "enabled": false,
            "groupName": "all-major-python-packages",
            "groupSlug": "all-major-python-packages",
            "matchDatasources": ["python"],
            "matchManagers": ["pipenv"],
            "matchUpdateTypes": ["major"]
          },
          {
            "enabled": true,
            "groupName": "python-packages",
            "groupSlug": "python-packages",
            "matchDatasources": ["python"],
            "lockFileMaintenance": {"enabled": true},
            "matchDepTypes": ["packages"],
            "matchManagers": ["pipenv"],
            "matchPackageNames": ["protobuf"],
            "matchUpdateTypes": ["minor", "patch"]
          }
...
WARN: findIssue() is not implemented (repository=test/renovate-test-repo)
INFO: Dependency extraction complete (repository=test/renovate-test-repo)
      "baseBranch": "master",
      "stats": {
        "managers": {"pipenv": {"fileCount": 1, "depCount": 1}},
        "total": {"fileCount": 1, "depCount": 1}
      }
(node:8) Warning: Setting the NODE_TLS_REJECT_UNAUTHORIZED environment variable to '0' makes TLS connections and HTTPS requests insecure by disabling certificate verification.
(Use `node --trace-warnings ...` to show where the warning was created)
INFO: Branch created (repository=test/renovate-test-repo, branch=renovate/protobuf-3.x)
      "commitSha": "14bfc87dc466804f0e831dbbfb6eee408936711f"
INFO: PR created (repository=test/renovate-test-repo, branch=renovate/protobuf-3.x)
      "pr": 198718,
      "prTitle": "chore(deps): update dependency protobuf to v3.20.1"
INFO: Branch created (repository=test/renovate-test-repo, branch=renovate/protobuf-4.x)
      "commitSha": "bff81e023286eeacaffd6e4644f55c099dbc1939"
INFO: PR created (repository=test/renovate-test-repo, branch=renovate/protobuf-4.x)
      "pr": 198719,
      "prTitle": "chore(deps): update dependency protobuf to v4"
WARN: ensureIssue() is not implemented (repository=test/renovate-test-repo)
INFO: Repository finished (repository=test/renovate-test-repo)
      "durationMs": 21047
##[section]Finishing: Run Renovate Bot on repos
```

Creates a PR for v3.20.1 and a PR for v4, even though major version is supposedly disabled?

**Reply by rarkins** [reply]
> ```
> {
>   "$schema": "https://docs.renovatebot.com/renovate-schema.json",
>   "description": "Python dependencies and Docker images",
>   "packageRules": [
>     {
>       "enabled": false,
>       "groupName": "all-major-python-packages",
>       "groupSlug": "all-major-python-packages",

groupSlug is not necessary - remove this

>       "matchDatasources": ["python"],
>       "matchManagers": ["pipenv"],
>       "matchUpdateTypes": [
>         "major"
>       ]
>     },

This rule says to combine all major updates into a group, but also to disable them. It's a bit counterintuitive.

>     {
>       "enabled": true,

The enabled=true here is not needed because packages matching this rule would already be enabled

>       "groupName": "python-packages",
>       "groupSlug": "python-packages",
>       "matchDatasources": ["python"],

Wrong, should be pypi, not python

>       "lockFileMaintenance": {
>         "enabled": true
>       },

Never put lockFileMaintenance inside a package Rule. The above 3 lines are wrong

>       "matchDepTypes": [
>         "packages"
>       ],
>       "matchManagers": ["pipenv"],
>       "matchPackageNames": [
>         "protobuf"
>       ],

You're defining a group here but only matching a narrow range (e.g. of size one in this example) of packages.


>       "matchUpdateTypes": [
>         "minor",
>         "patch"
>       ]
>     }
>   ]
> }
> ```


> 
> `renovate.config.js`: Then we target this new repo in `renovate.config.js` in the renovate build repository
> 
> ```
> // renovate config file
> 
> module.exports = {
> 		timezone: 'Europe/Copenhagen',
> 		extends: [
> 			":dependencyDashboard",
> 			":semanticPrefixFixDepsChoreOthers",
> 			":ignoreModulesAndTests",
> 			":autodetectPinVersions",
> 			":prHourlyLimit2",
> 			":prConcurrentLimit10",
> 			"group:monorepos",
> 			"group:recommended",
> 			"workarounds:all",
> 			":prHourlyLimitNone"
> 		],
> 		azureWorkItemId: 123456,
> 		endpoint: 'http://tfs/tfs/DefaultCollection',
> 
> 		platform: 'azure',
> 		onboardingConfig: {
> 			extends: ['config:base'],
> 		},

It's not a good idea to define both an `extends` in your global config but also have a different onboardingConfig.extends here. What were you trying to achieve?

> 
> 		repositories: [
> 			'test/renovate-test-repo'
> 		],
> 
> 		allowCustomCrateRegistries: true,
> 		allowScripts: true,
> 		exposeAllEnv: true,
> 		dryRun: false,
> 		printConfig: true,
> 
> 		"python": {
> 			"registryUrls": ["https://artifacts.sys.dom/artifactory/api/pypi/python/simple/"]
> 		}

Better to have this as a package rule with matchDatasources=pypi

> };
> ```
> 
> Then running the Renovate Bot build, the relevant output is:
> 
> ```
> ...
>           {
>             "enabled": false,
>             "groupName": "all-major-python-packages",
>             "groupSlug": "all-major-python-packages",
>             "matchDatasources": ["python"],
>             "matchManagers": ["pipenv"],
>             "matchUpdateTypes": ["major"]
>           },
>           {
>             "enabled": true,
>             "groupName": "python-packages",
>             "groupSlug": "python-packages",
>             "matchDatasources": ["python"],
>             "lockFileMaintenance": {"enabled": true},
>             "matchDepTypes": ["packages"],
>             "matchManagers": ["pipenv"],
>             "matchPackageNames": ["protobuf"],
>             "matchUpdateTypes": ["minor", "patch"]
>           }
> ...
> WARN: findIssue() is not implemented (repository=test/renovate-test-repo)
> INFO: Dependency extraction complete (repository=test/renovate-test-repo)
>       "baseBranch": "master",
>       "stats": {
>         "managers": {"pipenv": {"fileCount": 1, "depCount": 1}},
>         "total": {"fileCount": 1, "depCount": 1}
>       }
> (node:8) Warning: Setting the NODE_TLS_REJECT_UNAUTHORIZED environment variable to '0' makes TLS connections and HTTPS requests insecure by disabling certificate verification.
> (Use `node --trace-warnings ...` to show where the warning was created)
> INFO: Branch created (repository=test/renovate-test-repo, branch=renovate/protobuf-3.x)
>       "commitSha": "14bfc87dc466804f0e831dbbfb6eee408936711f"
> INFO: PR created (repository=test/renovate-test-repo, branch=renovate/protobuf-3.x)
>       "pr": 198718,
>       "prTitle": "chore(deps): update dependency protobuf to v3.20.1"
> INFO: Branch created (repository=test/renovate-test-repo, branch=renovate/protobuf-4.x)
>       "commitSha": "bff81e023286eeacaffd6e4644f55c099dbc1939"
> INFO: PR created (repository=test/renovate-test-repo, branch=renovate/protobuf-4.x)
>       "pr": 198719,
>       "prTitle": "chore(deps): update dependency protobuf to v4"
> WARN: ensureIssue() is not implemented (repository=test/renovate-test-repo)
> INFO: Repository finished (repository=test/renovate-test-repo)
>       "durationMs": 21047
> ##[section]Finishing: Run Renovate Bot on repos
> ```
> 
> Creates a PR for v3.20.1 and a PR for v4, even though major version is supposedly disabled?

You can see that the groupName you've applied is not being set, which indicates that your package rule is not matching. It's because of matchDatasources=python. It should be pypi instead.

**Reply by Atheuz** [reply]
OK. I imagine my problem was then the matchDatasources=python thing, that should instead be pypi. I'll try to apply this and the other things you've pointed out. Thank you so much for your help @rarkins!
